### DISCUSSION TITLE
Specifying custom types for request? (Typescript)

### DISCUSSION THREAD

**Discussion by danthegoodman1** [fastify/fastify#3247] [Q&A] [answered]
If I wanted to (without writing a plugin) add something onto the `FastifyRequest` or `FastifyResponse` types (as demnstrated below), how would I do so without having to specify a type overload every time like `(request as SomeExtendedType, response as AnotherExtendedType)`.

Also enabling me to do actions in hooks like custom JWT decoding and setting `request.decodedToken` for the rest of the handler to access for example.

```js
// This request and response object
server.get('/', async (request, response) => { ...
console.log(request.decodedToken)
```

Like how (for example) `fastify-jwt` adds a `request.user` field found here: https://github.com/fastify/fastify-jwt/blob/f726c1f553fad4de52d75d5c6f51638301029b15/index.d.ts#L58

**Comment by RafaelGSS** [comment, marked as answer]
You can use `FastifyRequest<{ Body: YourInterfaceHere }>` for body params. Further details: https://github.com/fastify/fastify/blob/main/types/request.d.ts#L6

**Reply by danthegoodman1** [reply]
I was talking more about the request object itself, and adding something to the request object, not the body object. So adding request.blah instead of request.body.blah

**Reply by RafaelGSS** [reply]
Oh ok, you should use the [declaration-merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html)

```ts
declare module 'fastify' {
  interface FastifyRequest {
    blah: string
  }
}
```

See: https://www.fastify.io/docs/latest/TypeScript/#creating-a-typescript-fastify-plugin

It's not specific to plugins, you can use it directly in your application as well.

**Reply by danthegoodman1** [reply]
I had tried that before, but maybe I did something wrong. Working now!

**Reply by AnthonyLzq** [reply]
Hi sir, for me it is not working, or maybe I am doing something wrong, but I'm getting the following error:

```sh
      Type 'RouteGenericInterface' is not assignable to type '{ Body: Test; }'.
        Types of property 'Body' are incompatible.
          Type 'unknown' is not assignable to type 'Test'
```

This is my code:

```ts
import { FastifyRequest } from 'fastify'
import { newResponse } from 'network/response'

interface DtoUser {
  id?: string
  lastName?: string
  name?: string
}

type CustomRouter = (
  prefix?: string
) => import('fastify').RouteOptions | import('fastify').RouteOptions[]

type Test = { args: DtoUser }

const User: CustomRouter = (prefix = '/user') => [
  {
    url: `${prefix}`,
    method: 'GET',
    handler: (request: FastifyRequest<{ Body: Test }>, reply) => {
      const {
        body: { args }
      } = request
      newResponse({
        error: false,
        message: 'User',
        reply,
        status: 200
      })
    }
  }
]
```

**Reply by kriskw1999** [reply]
If you declare the intersection in a module where the fastify is not imported you may override fastify types at all. To fix this try importing fastify at the beginning of the module like:
```ts
import "fastify";

declare module "fastify" {
  interface FastifyRequest {
    authenticate: () => Promise<any>;
  }
}

```

In this way you will merge the interface instead of overriding it keeping the generic and so on.
